regtests/test_mbtr_derivatives.py

```diff
@@ -0,0 +1,138 @@
+        d["grid"] = {"min": 0, "max": 1.0, "sigma": 0.02, "n": 100}
+    else:
+        d["geometry"] = {"function": "distance"}
+        d["grid"] = {"min": 0, "max": 10.0, "sigma": 0.2, "n": 100}
+
+    if weighting_function == "exp":
+        d["weighting"] = {"function": "exp", "r_cut": 10.0, "threshold": 1e-3}
+
+    return d
+
+
+def k3_dict(weighting_function):
+    d = {}
+    d["geometry"] = {"function": "cosine"}
+    d["grid"] = {"min": -1.0, "max": 1.0, "sigma": 0.02, "n": 100}
+
+    if weighting_function == "exp":
+        d["weighting"] = {"function": "exp", "scale": 1.0, "threshold": 1e-3}
+
+    return d
+
+
+def mbtr(**kwargs):
+    """Returns a function that can be used to create a valid MBTR
+    descriptor for a dataset.
+    """
+
+    def func(systems=None):
+        species = list(set().union(*[s.get_atomic_numbers() for s in systems]))
+
+        final_kwargs = {
+            "species": species,
+            "normalization": "none",
+            "periodic": True,
+            "sparse": False,
+            "flatten": True,
+            "k1": {
+                "geometry": {"function": "atomic_number"},
+                "grid": {"min": 0, "max": 9, "sigma": 0.1, "n": 100},
+            },
+            "k2": k2_dict("inverse_distance", "exp"),
+            "k3": k3_dict("exp")
+        }
+        final_kwargs.update(kwargs)
+
+        return MBTR(**final_kwargs)
+
+    return func
+
+
+def assert_derivatives_analytical(descriptor_func):
+    """Test analytical values against a numerical python implementation."""
+    system = water()
+    system.pbc = [True, True, True]
+    h = 0.0001
+    n_atoms = len(system)
+    n_comp = 3
+    descriptor = descriptor_func([system])
+
+    n_features = descriptor.get_number_of_features()
+    derivatives_python = np.zeros((n_atoms, n_comp, n_features))
+    d0 = descriptor.create(system)
+    coeffs = [-1.0 / 2.0, 1.0 / 2.0]
+    deltas = [-1.0, 1.0]
+    for i_atom in range(len(system)):
+        for i_comp in range(3):
+            for i_stencil in range(2):
+                system_disturbed = system.copy()
+                i_pos = system_disturbed.get_positions()
+                i_pos[i_atom, i_comp] += h * deltas[i_stencil]
+                system_disturbed.set_positions(i_pos)
+                d1 = descriptor.create(system_disturbed)
+                derivatives_python[i_atom, i_comp, :] += coeffs[i_stencil] * d1 / h
+
+    # Calculate analytical derivatives
+    derivatives_analytical, d_cpp = descriptor.derivatives(system, method="analytical")
+
+    # Compare descriptor values
+    assert np.allclose(d0, d_cpp, atol=1e-6)
+
+    # Compare derivative values. The check is not very strict because there
+    # is no guarantee that the numerical derivatives are accurate
+    max_error = np.max(np.abs(derivatives_python - derivatives_analytical))
+    max_value = np.max(np.abs(derivatives_python))
+    assert max_error / max_value < 1e-2
```

Comment:
Your comment about floating point accuracy is exactly what I meant. 32-bit accuracy does not allow using a low enough h-value, which makes numerical derivatives inaccurate. 32-bit accuracy is not a problem for the analytical derivative calculations any more than it is for the descriptor calculations.

I do, however, think that changing to 64-bit is a good idea. I made the migration and implemented a `dtype`-option mimicking SOAP. There were no measurable changes to computation time.

Even after this change I had problems reaching the desired accuracy. It turns out that in the tests I had set the MBTR cutoff to be exactly twice the simulation cell width, which means that periodic images of all atoms were sitting exactly at the cutoff. Because the representation function is not continuous at the cutoff, the numerical and analytical derivatives were not the same. Adjusting the cutoff radius solved the problem.

Although I agree that using same code for all the tests is a good idea, I did not make that change for now. `assert_derivatives_numerical`-function uses `big_system` for the tests. Due to the symmetries of this system, the derivatives of k2-term are 0 for all atoms at least up to a cutoff of 6 angstroms. With larger cutoff values, calculating numerical derivatives becomes prohibitively slow. Calculating k3-derivatives (especially when weighting is turned off) is even slower. The tests with the water system allow more complete testing of different MBTR settings.